Clean Encoder-Decoder models with Bart/T5-like API and add generate possibility

patrickvonplaten opened the pull request:
Bert-Bert Encoder-Decoder models can now be used as is shown in the test cases: 
`tests/test_modeling_encoder_decoder.py`. 
Tests include:
- forward `input_ids` and `decoder_input_ids` for Bert-Bert
- backprop using masked language model loss for Bert-Bert
- backprop using "conventional" language model loss for Bert-Bert
- using the `generate()` fn with Bart-Bart
- saving and loading of Encoder-Decoder models.

Before merging a couple of things have to be agreed on as mentioned in the comments below.

UPDATE: 

This branch is IMO now fully functional for Bert-2-Bert models. 
I will finish the PR (clean the code, make a pretty docstring, etc...) once we agreed on the issues I mentioned further down. Would be very happy if you can review @thomwolf @LysandreJik @sshleifer @julien-c @yjernite

mgoldey commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -201,36 +218,65 @@ def save_pretrained(self, save_directory):
+
+        # Decode
+        decoder_outputs = self.decoder(
+            input_ids=decoder_input_ids,
```
I remain confused about the input here since it seems incompatible with inference to train it with the decoder_input_ids. Can you elucidate?

mgoldey replied:
Ah, this might be a use case confusion on my part. I'm envisioning seq2seq, but I think this is a generative model.

mgoldey commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -18,35 +18,58 @@
     def __init__(self, encoder, decoder):
-        super().__init__()
+        assert encoder is not None, "The encoder has to be defined"
+        assert decoder is not None, "The encoder has to be defined"
```
```suggestion
        assert decoder is not None, "The decoder has to be defined"
```

patrickvonplaten commented on src/transformers/modeling_bert.py:
```diff
@@ -957,7 +957,29 @@ def forward(
-        return outputs  # (masked_lm_loss), (ltr_lm_loss), prediction_scores, (hidden_states), (attentions)
+        return outputs  # (ltr_lm_loss), (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
+
+    def prepare_inputs_for_generation(self, input_ids, attention_mask=None, **model_kwargs):
```
Since Bert is usually pretrained with masked language modeling, for generation a new "dummy" token is added, on which we put a mask so that the next token can be predicted by Bert. This is similar to how it is done for XLNet e.g.: https://github.com/huggingface/transformers/blob/857ccdb259b7e46c60cf86c58b7ab038c63e4d4e/src/transformers/modeling_xlnet.py#L941

patrickvonplaten commented on src/transformers/modeling_bert.py:
```diff
@@ -957,7 +957,29 @@ def forward(
+            (effective_batch_size, 1), self.config.pad_token_id, dtype=torch.long, device=input_ids.device
+        )
+        input_ids = torch.cat([input_ids, dummy_token], dim=1)
+
```
The `attention_mask` logic is quite important. The `prepare_inputs_for_generation()` is set up for two cases:
1) Bert is used as a stand-alone decoder. In this case the "correct" `attention_mask` is given by the `generate()` function and will be extended by one
2) Bert is used as a encoder-decoder model (Bert-2-Bert). In this case this function is called for the decoder Bert only and for which we have to extend the `decoder_attention_mask`, which is **not** given (yet) by the `generate()` function. Therefore we have to create it here.

patrickvonplaten commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -18,35 +18,56 @@
         self.decoder = decoder
 
+    def _init_config(self, encoder_config, decoder_config):
+        # decoder config is used as default config (important for generation)
```
The class EncoderDecoderModel needs a config to be able to instantiate a PretrainedModel to be able to use the `generate()` function. The question is which config we use here. For the moment I just use the given config of the decoder. I think we should maybe also add a couple of assert statements to check that the configs are compatible here, but no sure what to check exactly. What do you think @sshleifer @thomwolf @LysandreJik

patrickvonplaten commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -116,38 +137,29 @@ def from_pretrained(
+            assert (
+                decoder_pretrained_model_name_or_path is not None
+            ), "If `decoder_model` is not defined as an argument, a `decoder_pretrained_model_name_or_path` has to be defined"
             decoder = AutoModelWithLMHead.from_pretrained(decoder_pretrained_model_name_or_path, **kwargs_decoder)
```
At the moment we allow decoder models only to be models with an LMHead - we could make it more flexible by added a flag argument to the `from_pretrained()` function. What do you think @thomwolf @LysandreJik @sshleifer ?

patrickvonplaten commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -201,36 +213,70 @@ def save_pretrained(self, save_directory):
-        `encoder_hidden_state` is passed to this function.
+    def forward(
+        self,
+        input_ids=None,
```
The API is very similar to what is done in `T5` and `Bart`. I also thought about just having a single `**kwargs` function argument that is then split into `decoder_kwargs` and `encoder_kwargs`. These kwargs can then just be inserted into their respective models. This would make the function a bit more flexible (mabye people want to use special Bert arguments for `next_sentence_prediction`, etc.), but less readable. A compromise, would maybe be to leave it as it is now and additionally add the **kwargs argument that is split in put into the correct models. What do you think @sshleifer @thomwolf @LysandreJik @yjernite ?

patrickvonplaten commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -201,36 +213,70 @@ def save_pretrained(self, save_directory):
+            "encoder_outputs": encoder_outputs,
+        }
+
+    def _reorder_cache(self, past, beam_idx):
```
This might need to be updated in the future

patrickvonplaten commented on src/transformers/utils_encoder_decoder.py:
```diff
@@ -1,47 +0,0 @@
-# coding=utf-8
```
Useless file IMO

patrickvonplaten commented on tests/test_modeling_encoder_decoder.py:
```diff
@@ -0,0 +1,177 @@
+
+# this line reruns all the tests in BertModelTest; not sure whether this can be prevented
+# for now only run module with pytest tests/test_modeling_encoder_decoder.py::EncoderDecoderModelTest
+from .test_modeling_bert import BertModelTest
```
This was quite annoying. As far as I can see it is not possible to import a unittest testcase, such as `BertModelTest from .test_modeling_bert` without running all of its test. At the same time I think we need to import the TestCase here since we need to instantiate BertEncoder and BertDecoder models without copy pasting all of its code. I tried a couple of hacks, but didn't find a good solution. Maybe it's possible to only run `pytest tests/test_modeling_encoder_decoder.py::EncoderDecoderModelTest` from cicrle ci in which case only the encoder-decoder model tests are run? Any good ideas for this case? @julien-c @LysandreJik @thomwolf @sshleifer

sshleifer replied:
1) Move the ModelTester to the top level of that file.
2) in that file, `self.model_tester = BertModelTester(self)`

3) only import `BertModelTester`

verified that no extra tests are run.

You can merge https://github.com/huggingface/transformers/pull/4027 if you want to keep your delta small.

patrickvonplaten replied:
Awesome! I will try this out :-)

patrickvonplaten replied:
Would it be ok to change Bert's testing structure as shown in PR #4027 for you guys @julien-c @thomwolf @LysandreJik ?

patrickvonplaten commented on src/transformers/modeling_bert.py:
```diff
@@ -957,7 +957,30 @@ def forward(
+        return outputs  # (ltr_lm_loss), (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
+
+    def prepare_inputs_for_generation(self, input_ids, attention_mask=None, **model_kwargs):
+        # Add dummy token at the end (no attention on this one)
```
I added now two cases here:
1) Bert is a decoder and therefore does casual masking. For this case, I assume it has been trained using the "normal" language modeling loss. Thus no need to add a dummy token.
2) Bert is **not** a decoder and therefore was trained using masked language modeling and needs to add a dummy token similar to how it's done for XLNet: https://github.com/huggingface/transformers/blob/778cf7cff6ffd32281b7583a0db0b566c0c76aee/src/transformers/modeling_xlnet.py#L941

What do you think @LysandreJik @thomwolf @sshleifer @yjernite ?

thomwolf commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -18,35 +18,57 @@
+        class method for the encoder and `AutoModelWithLMHead.from_pretrained(pretrained_model_name_or_path)` class method for the decoder.
     """
 
     def __init__(self, encoder, decoder):
```
So as I detailed below under the `from_pretrained` method, I think here we should allow both an initialization from (pretained) sub-parts and from a config so we can also just reload a saved pretrained model.

Please read the long comment below `from_pretrained` before this one.

Here is a proposal for the `__init__` that should allow both loading:
```python
    def __init__(self, config=None, encoder=None, decoder=None):
        assert config is not None or (encoder is not None and decoder is not None), "Either a configuration or an Encoder and a decoder has to be provided"

        if config is not None:
            self.config = config
            self.encoder = AutoModel.from_config(config.encoder)
            self.decoder = AutoModelWithLMHead.from_config(config.decoder)
        else:
            self.config = EncoderDecoderConfig.from_encoder_decoder_config(encoder.config, decoder.config)
            self.encoder = encoder
            assert (
                self.encoder.get_output_embeddings() is None
            ), "The encoder {} should not have a LM Head. Please use a model without LM Head"
            self.decoder = decoder
```

thomwolf replied:
This mean we need an `EncoderDecoderConfig` which will basically be the following:

```python
class EncoderDecoderConfig(PretrainedConfig):
    model_type = "encoder_decoder"
    def __init__(self, **kwargs):
        encoder_config = kwargs.pop('encoder')
        encoder_model_type = encoder_config.pop('model_type')
        decoder_config = kwargs.pop('decoder')
        decoder_model_type = decoder_config.pop('model_type')
        self.encoder = AutoConfig.for_model(encoder_model_type, **encoder_config)
        self.decoder = AutoConfig.for_model(decoder_model_type, **decoder_config)

    @classmethod
    def from_encoder_decoder_config(cls, encoder_config, decoder_config):
        return cls(encoder=encoder_config.to_dict(), decoder=decoder_config.to_dict())

     def to_dict(self):
        """
        Serializes this instance to a Python dictionary. Override the default `to_dict()` from `PretrainedConfig`.

        Returns:
            :obj:`Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance,
        """
        output = copy.deepcopy(self.__dict__)
        output['encoder'] = self.encoder.to_dict()
        output['decoder'] = self.decoder.to_dict()
        output["model_type"] = self.__class__.model_type
        return output
```

sshleifer commented on src/transformers/modeling_bert.py:
```diff
@@ -957,7 +957,29 @@ def forward(
+        if attention_mask is None:
+            attention_mask = input_ids.new_ones(input_shape)
+
+        # if model is does not use a casaul mask then add a dummy token
```
typo causal

sshleifer commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -18,35 +18,57 @@
-        decoder_pretrained_model_name_or_path=None,
-        *model_args,
-        **kwargs
+        cls, pretrained_model_name_or_path=None, decoder_pretrained_model_name_or_path=None, *model_args, **kwargs
```
I like this a lot: both should be possible.
We could simplify the code path by having normal from_pretrained and a second classmethod:

```python
def from_pretrained_encoder_decoder_(encoder_pretrained_model_name_or_path=None, decoder_pretrained_model_name_or_path=None):

```

That way `from_pretrained` is consistent in signature and function with the rest of the lib, but the extra functionality is still possible.

patrickvonplaten commented on src/transformers/configuration_encoder_decoder.py:
```diff
@@ -0,0 +1,79 @@
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        assert "encoder" in kwargs and "decoder" in kwargs, "Config has to be initialized with encoder and decoder config"
```
@thomwolf  I added this assert statement because I don't really see a reason to not have a encoder and decoder config.

patrickvonplaten commented on src/transformers/configuration_encoder_decoder.py:
```diff
@@ -0,0 +1,79 @@
+        self.is_encoder_decoder = True
+
+    @classmethod
+    def from_encoder_decoder_config(cls, encoder_config, decoder_config):
```
Note that EncoderDecoderConfig can be loaded via both:
- `from_encoder_decoder_pretrained` and `from_pretrained` (as is the case for the model)

patrickvonplaten commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -16,37 +16,65 @@
-        decoder_pretrained_model_name_or_path=None,
-        *model_args,
-        **kwargs
+    def from_encoder_decoder_pretrained(
```
@thomwolf  As @sshleifer suggested I changed the name here to be consistent with the config name. In order to load a "complete" encoder-decoder model the normal `from_pretrained` fn can be used. 

Note that `EncoderDecoderModel` can be loaded via both:

`from_encoder_decoder_pretrained` and `from_pretrained` (as is the case for the config)

thomwolf replied:
Ok!

patrickvonplaten commented:
UPDATE:

I really liked the idea of adding a encoder_decoder config (@thomwolf ) and having for both the encoder_decoder config and model 2 `from_pretrained` fn:
1. The standard one which is used thanks to inheritence to `PretrainedConfig` and `PretrainedModel` 
2. a `from_encoder_decoder_pretrained` fn for both config and model (@sshleifer)

To understand how to use the encoder decoder class please confer to the added tests.

sshleifer commented on src/transformers/configuration_encoder_decoder.py:
```diff
@@ -0,0 +1,82 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" BERT model configuration """
```
typo

sshleifer commented on src/transformers/configuration_encoder_decoder.py:
```diff
@@ -0,0 +1,82 @@
+# coding=utf-8
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
```
typo

sshleifer commented on src/transformers/configuration_encoder_decoder.py:
```diff
@@ -0,0 +1,82 @@
+        self.is_encoder_decoder = True
+
+    @classmethod
+    def from_encoder_decoder_pretrained(cls, encoder_config, decoder_config):
```
`<nit>`

the other from_pretrained takes `pretrained_model_name_or_path: str`, this takes 2 configs.
I would name it `from_configs`...
From the name you might also guess the arguments are instances of models.
Also type hints would be nice here.
`</nit>`

sshleifer replied:
I understand the consistency argument though. Fair enough.

patrickvonplaten replied:
Hmm, I see your point! For me it would be fine to rename it to `from_encoder_decoder_configs()` What do you think @thomwolf @LysandreJik @julien-c ?

thomwolf replied:
I'm neutral. Choose the option you like @patrickvonplaten

patrickvonplaten replied:
Ok went for @sshleifer suggestion.

sshleifer commented on src/transformers/modeling_bert.py:
```diff
@@ -959,6 +959,28 @@ def forward(
+        input_shape = input_ids.shape
+        effective_batch_size = input_shape[0]
+
+        # if model is used as a decoder in encoder-decoder model decoder attention mask is created on the fly
```
comma between model and decoder

sshleifer commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -16,32 +16,68 @@
     """
+    config_class = EncoderDecoderConfig
+
+    def __init__(self, config=None, encoder=None, decoder=None):
```
type hints would be helpful

sshleifer commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -108,129 +118,116 @@ def from_pretrained(
-            encoder_hidden_states = encoder_outputs[0]
-        else:
-            encoder_outputs = ()
+        kwargs_encoder = {argument: value for argument, value in kwargs.items() if not argument.startswith("decoder_")}
```
so neutral kwargs like use_cache only go to encoder? Is that desired/irrelevant?

patrickvonplaten replied:
I think the user will always have to prepend each argument with `decoder_` in order to use it for the encoder. So if one would use a Bert + GPT2 (which is not supported at the moment), one would have to write `decoder_use_cache=True`. IMO this is ok.

sshleifer commented on src/transformers/modeling_encoder_decoder.py:
```diff
@@ -108,129 +118,116 @@ def from_pretrained(
+
+    def _reorder_cache(self, past, beam_idx):
+        # as a default encoder-decoder models do not re-order the past.
+        # TODO(PVP): might have to be updated, e.g. if GPT2 is to be used as a decoder
```
```python
tuple(layer_past.index_select(1, beam_idx) for layer_past in past)
```
This is a better IMO or `raise NotImplementedError` isn't this implem incorrect?

patrickvonplaten replied:
It should not raise anything because the function is called in `generate` and should not reorder anything. I think for the moment this is ok.

sshleifer commented on tests/test_modeling_encoder_decoder.py:
```diff
@@ -0,0 +1,333 @@
+# coding=utf-8
+# Copyright 2018 Google T5 Authors and HuggingFace Inc. team.
```
typo

sshleifer commented on tests/test_modeling_encoder_decoder.py:
```diff
@@ -0,0 +1,333 @@
+        input_ids_dict = self.prepare_config_and_inputs_bert()
+        self.create_and_check_save_and_load(**input_ids_dict)
+
+    def test_save_and_load_from_encoder_decoder_pretrained(self):
```
I don't understand the point of all these helper functions that you use once.

patrickvonplaten replied:
Do you mean the `create_and_check...` functions? It's done everywhere in the tests and I think it's actually a nice way to have short and easy to read/see test functions at the bottom and all the logic further up

patrickvonplaten commented:
Code is cleaned: added type hints, cleaned the docstring and added a encoder-decoder model page.
Just need to resolve the issue with importing Bert's model tester. @sshleifer found a solution. If everybody is fine with it - I'll go for it :-)

sshleifer commented:
LGTM!

patrickvonplaten commented:
Ok Good to merge for me! If @sshleifer it's ok for you I will use your PR #4027 for the new test proposition.

sshleifer commented:
Yes!
